libuser/src/threads.rs

```diff
@@ -59,15 +59,71 @@
 use crate::types::Thread as ThreadHandle;
 use crate::syscalls;
 use crate::error::Error;
+use crate::error::KernelError;
 use crate::thread_local_storage::TlsElf;
 use sunrise_libkern::{TLS, IpcBuffer};
 use alloc::boxed::Box;
+use alloc::alloc::{alloc, dealloc, Layout};
 use core::mem::ManuallyDrop;
 use core::fmt;
 use spin::Once;
 
-/// Size of a thread's stack, in bytes.
-const STACK_SIZE: usize = 0x8000;
+/// Default size of a thread's stack, in bytes.
+pub const DEFAULT_STACK_SIZE: usize = 0x8000;
+
+/// Stack allocation informations
+#[derive(Debug)]
+struct StackContext {
+    /// The addresss of the allocated stack
+    stack_address: *const u8,
+
+    /// The stack layout.
+    stack_layout: Layout
+}
+
+impl StackContext {
+    /// Create a new StackContext from a given size. The stack size must be bigger than 0.
+    ///
+    /// # Errors
+    ///
+    /// - `InvalidSize`
+    ///   - The size passed was 0
+    ///   - The size overflows when rounded up to the nearest multiple of PAGE_SIZE.
+    pub fn new(stack_size: usize) -> Result<Self, Error> {
+        if stack_size == 0 {
+            return Err(KernelError::InvalidSize.into());
+        }
+
+        let stack_layout = Layout::from_size_align(stack_size, crate::mem::PAGE_SIZE)
+            .or(Err(KernelError::InvalidSize))?;
+
+        Ok(StackContext {
+            stack_address: unsafe {
+                // Safety: We manually alloc the stack from the given stack_layout and dealloc it on Drop.
```

Comment:
This isn't a good safety comment. You need to explain why this piece of unsafe code is sound. The alloc function [documents the situations in which it is unsound](https://doc.rust-lang.org/alloc/alloc/trait.GlobalAlloc.html#safety-1). We need to show that we never hit those cases.

```suggestion
                // Safety: We error from the function early if stack_size is 0. We don't care much about whether the block is initialized.
```